### ISSUE TITLE
Use OpenConnect With A GUI Application On Linux

### ISSUE THREAD

**Issue by davidpgil** [dlenski/openconnect#63]
I have managed to get OpenConnect working for my situation with the help of this thread -- [How to use OpenConnect without specifying a certificate?](https://github.com/dlenski/openconnect/issues/60) ... Now the next step for me ideally would be to not have to type in all of this via CLI every time...

I use Trisquel GNU+Linux 8.0 (Ubuntu/Debian derived). I installed _network-manager-openconnect-gnome_ in order to add OpenConnect support to Network Manager. Howver, I am not sure how to use the OpenConnect in this Github repository, with the Network Manager GUI for OpenConnect that I see in the following screenshot.

![screenshot at 2017-11-05 09 21 26](https://user-images.githubusercontent.com/1891288/32415649-f1482ff4-c20a-11e7-8bde-ab7d416a37a9.png)

It seems I need a certificate? I tried using _gnutls --print-cert_ to generate a certificate, but I do not understand enough about what I am doing to get it to work. Please advise.

**Comment by dlenski** [comment]
Short answer: Don't. Use the CLI interface for now, and create a shell script as a shortcut for the appropriate connection parameters. 

Long answer:

1. You need to build the NM plugin from source as well. The released versions of the plugin included work Linux distros don't include multiprotovol support. Unreleased version of NM plugin that does support GlobalProtect is here: https://github.com/dlenski/network-manager-openconnect
2. I don't know what you mean by "need a certificate". OpenConnect uses certificates to authenticate VPN servers, and optionally as client authentication, but the latter is only necessary if the server requires it. Based on your previous issue report, I think you're just getting confused by the fact that the NM plugin you're using *doesn't know how to connect to GlobalProtect VPNs at all*.

**Comment by davidpgil** [comment]
I like #1. Thanks.

Regarding #2, what I mean by "need a certificate" is that the NM GUI throws an error if I don't give it a PEM file. The NM applet i'm using has an entry in the Connection Type pull down list that says "Cisco AnyConnect Compatible VPN (openconnect)" so logically, I assumed that it would be compatible with your CLI app named "openconnect".

**Comment by dlenski** [comment]
> … the NM GUI throws an error if I don't give it a PEM file. 

Hmmm. Really? In which of the certificate fields? The "CA Certificate" or the "User certificate"?

It should *not* require a user certificate unless your VPN requires a client user SSL certificate. As for the CA certificate… if the server certificate can't be verified it should pop up a prompt asking if you want to approve the unverified server certificate, rather than throw an error.

> The NM applet i'm using has an entry in the Connection Type pull down list that says "Cisco AnyConnect Compatible VPN (openconnect)", so logically, I assumed that it would be compatible with your CLI app named "openconnect".

Yeah, it is confusing. Unfortunately, the officially released versions of `openconnect` don't provide any way for GUIs to figure out which clients are supported.

Making this work requires a coordinated patch to both the base openconnect ([already in the official repository](http://git.infradead.org/users/dwmw2/openconnect.git/commitdiff/40802e459873e55eefbe64f05b9475821f761b8e) but not yet released) and to the NM plugin ([ditto](https://git.gnome.org/browse/network-manager-openconnect/commit/?id=1ed6de801a398887277fbb9c35d9f5f5d6c4d5c3)).

Once GlobalProtect support is merged into the mainline and new versions of both openconnect and the NM plugin are released… we'll all live happily ever after :joy:

**Comment by davidpgil** [comment]
I will just share what I see in the sequence I see it. After clicking on "Edit Connections..." on the Network Manager, I see the following...

![screenshot at 2017-11-08 05 06 03](https://user-images.githubusercontent.com/1891288/32543207-f5fff128-c442-11e7-97f5-b8d61ce7d2c7.png)

After clicking "Create..."

![screenshot at 2017-11-08 05 08 58](https://user-images.githubusercontent.com/1891288/32543265-21529808-c443-11e7-8d17-172d971e5969.png)

As you can see "✔ Save" is grayed out.

Have an ideas on what I could do to proceed? If not, it's fine to Archive the issue. I think anyone seeing this would be sufficiently helped as far as the current state of this issue at least. Thanks for your help thus far.

**Comment by dlenski** [comment]
> As you can see "✔ Save" is grayed out.
>
> Have an ideas on what I could do to proceed? If not, it's fine to Archive the issue. I think anyone seeing this would be sufficiently helped as far as the current state of this issue at least. Thanks for your help thus far.

I believe Save is only grayed-out because you haven't entered the VPN server address (e.g. `vpngateway.company.com`), which certainly **is** required :grinning:

**Comment by davidpgil** [comment]
Ha. I think I did try this already, but I will try it again in a bit, smartypants. 😅

**Comment by dlenski** [comment]
The certificate fields certainly should _not_ be required to be populated for any version of the GUI.

Here is what the GUI should look like once you've built and installed my version with the multi-protocol support:

![image](https://user-images.githubusercontent.com/128716/32553850-a2354382-c4a0-11e7-803b-f8b5d696d5e0.png)

**Comment by davidpgil** [comment]
I'll try this as soon as I am able, later today. Thanks.

**Comment by dlenski** [comment]
As I wrote above:

> The released versions of the plugin included with Linux distros don't include multiprotocol support

… so when you try to connect to a **GP** VPN using the released plugin, the plugin is trying to interact with the VPN server as if it were a **Cisco AnyConnect** VPN, not a GP VPN. Hence the error.

> On second thought, if you don't mind, would you give me a sense of how to build your version of the OpenConnect NM …

Sorry, but I don't have time to give specific build advice on it. It's more complex than building the CLI version. The code is "out there" mainly for review by developers and integrators, not for general-purpose use at this point.
